runtime/libtapasco/src/dma.rs

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2014-2020 Embedded Systems and Applications, TU Darmstadt.
+ *
+ * This file is part of TaPaSCo
+ * (see https://github.com/esa-tu-darmstadt/tapasco).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use crate::device::DeviceAddress;
+use crate::device::DeviceSize;
+use crate::tlkm::tlkm_copy_cmd_from;
+use crate::tlkm::tlkm_copy_cmd_to;
+use crate::tlkm::tlkm_ioctl_copy_from;
+use crate::tlkm::tlkm_ioctl_copy_to;
+use core::fmt::Debug;
+use memmap::MmapMut;
+use snafu::ResultExt;
+use std::fs::File;
+use std::os::unix::prelude::*;
+use std::sync::Arc;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Could not transfer to device {}", source))]
+    DMAToDevice { source: nix::Error },
+
+    #[snafu(display("Could not transfer from device {}", source))]
+    DMAFromDevice { source: nix::Error },
+
+    #[snafu(display(
+        "Transfer 0x{:x} - 0x{:x} outside of memory region 0x{:x}.",
+        ptr,
+        end,
+        size
+    ))]
+    OutOfRange {
+        ptr: DeviceAddress,
+        end: DeviceAddress,
+        size: DeviceSize,
+    },
+
+    #[snafu(display("Failed flushing the memory for DirectDMA: {}", source))]
+    FailedFlush { source: std::io::Error },
+}
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub trait DMAControl: Debug {
+    fn copy_to(&self, data: &[u8], ptr: DeviceAddress) -> Result<()>;
+    fn copy_from(&self, ptr: DeviceAddress, data: &mut [u8]) -> Result<()>;
+}
+
+#[derive(Debug, Getters)]
+pub struct DriverDMA {
```

Comment:
Again, some comments regarding the difference between DriverDMA and DirectDMA would be good